src/transport_performance/analyse_network.py

```diff
@@ -0,0 +1,441 @@
+            sel_origins = origins[o : o + num_origins]
+
+            # calculates cross join of origins and destinations
+            full_gdf = orig_gdf[orig_gdf["id"].isin(sel_origins)].merge(
+                dest_gdf, how="cross", suffixes=["_orig", "_dest"]
+            )
+
+            # calculates haversine distance between origins and destinations
+            full_gdf["distance"] = self._haversine_df(
+                full_gdf,
+                f"{geometry_col}_orig",
+                f"{geometry_col}_dest",
+                unit=unit,
+            )
+
+            # filters out pairs where distance is over threshold
+            sel_dest = list(
+                full_gdf[full_gdf["distance"] < distance]["id_dest"].unique()
+            )
+
+            # yields lists with selected origins and destinations
+            yield sel_origins, sel_dest
+
+    def _haversine_df(
+        self,
+        df: pd.DataFrame,
+        orig: str,
+        dest: str,
+        unit: Unit,
+    ) -> np.array:
+        """Calculate haversine distance between shapely Point objects.
+
+        Parameters
+        ----------
+        df : pd.DataFrame
+            Cartesian product of origin and destinations geodataframes. Should
+            contain a geometry for each.
+        orig : str
+            Name of the column containing the origin geometry.
+        dest : str
+            Name of the column containing the destination geometry.
+        unit : Unit
+            Unit to calculate distance.
+
+        Returns
+        -------
+        dist_array : np.array
+            Array with row-wise distances.
+
+        """
+        # defences
+        d._type_defence(df, "df", pd.DataFrame)
+        d._check_column_in_df(df, orig)
+        d._check_column_in_df(df, dest)
+        d._type_defence(unit, "unit", Unit)
+
+        lat_long_orig = df[orig].apply(lambda x: (x.y, x.x))
+        lat_long_dest = df[dest].apply(lambda x: (x.y, x.x))
+
+        dist_array = haversine_vector(
+            list(lat_long_orig), list(lat_long_dest), unit=unit
+        )
+
+        return dist_array
+
+    def _estimate_num_partitions(
+        self, df: pd.DataFrame, partition_size: int = 200
+    ) -> int:
+        """Estimate number of parquet partitions needed.
+
+        Parameters
+        ----------
+        df: pd.DataFrame
+            Dataframe to save as parquet file.
+        partition_size: int
+            Maximum size in MB for each parquet file partition. Defaults to
+            200 MB.
+
+        Returns
+        -------
+        int
+            Required partitions for the parquet file.
+
+        """
+        # defences
+        d._type_defence(df, "df", pd.DataFrame)
+        d._type_defence(partition_size, "partition_size", int)
+
+        mem_usage = sum(df.memory_usage(deep=True, index=False))
+        est_parts = ceil(mem_usage / (partition_size * 1e6))
+
+        return est_parts
+
+    def _save_to_parquet(
```

Comment:
Doing nothing may lead to some issues. If the number of files being produced is less than the number of files already in the folder (e.g. we're running all origins when we run single origins before), it will only overwrite some of the files leaving the others. This will cause duplication when loading the parquet files for the metrics.

Not sure what would be the best way to go. I think we should definitely empty the folder before saving new files, but not sure how the user should be made aware of this. Should we raise an error if the folder has content? Should we add a flag to empty the folder? Happy to discuss.

Reply:
I have added the defence in the `__init__`, which raises a `NotImplementedError` if files are in the provided output directory.

Tests have become a bit more complicated as a result. To create the `dummy_transport_network` fixture we need to create a module-scoped temp path. However, to test i/o behaviour we need to use a function-scoped temp path.

Reply:
Step 1. above has been implemented and is working exactly as expected - raises an Error when parquet files already exist in the output folder to prevent overwriting. This can be circumnavigated by calling `od_matrix` with different configurations after an `AnalyseNetwork` instance has been instantiated. However, this approach does not align with the design choice made to priorities error raising as soon as possible. Nor is it, at this time, the expected way to use this class. For these reasons, plus the fact the class if functioning as needed, this last point can be considered as tech-debt and resolved at a later date. Ticket #222 has been raised to capture this tech-debt. Closing the conversation with this ticket.